datalad/local/foreach.py

```diff
@@ -0,0 +1,275 @@
+"""Plumbing command for running a command on each (sub)dataset"""
+
+__docformat__ = 'restructuredtext'
+
+
+import logging
+
+from argparse import REMAINDER
+from itertools import chain
+from tempfile import mkdtemp
+
+from datalad.cmd import NoCapture, StdOutErrCapture
+from datalad.core.local.run import normalize_command
+
+from datalad.distribution.dataset import (
+    Dataset,
+    EnsureDataset,
+    datasetmethod,
+    require_dataset,
+)
+from datalad.dochelpers import exc_str
+from datalad.interface.base import (
+    Interface,
+    build_doc,
+)
+from datalad.interface.common_opts import (
+    jobs_opt,
+    recursion_flag,
+    recursion_limit,
+)
+from datalad.interface.results import get_status_dict
+from datalad.interface.utils import eval_results
+from datalad.support.constraints import (
+    EnsureBool,
+    EnsureNone,
+)
+from datalad.support.parallel import (
+    ProducerConsumer,
+    ProducerConsumerProgressLog,
+    no_parentds_in_futures,
+    no_subds_in_futures,
+)
+from datalad.support.param import Parameter
+from datalad.utils import (
+    SequenceFormatter,
+    getpwd,
+    shortened_repr,
+)
+
+lgr = logging.getLogger('datalad.local.foreach')
+
+
+@build_doc
+class ForEach(Interface):
+    r"""Run a command on the dataset and/or each of its sub-datasets.
+
+    Each dataset's repo runner is reused to execute the commands, so they
+    are are executed in the top directory of a corresponding dataset.
+    In contrast, Python commands (expressions) are evaluated without changing directory.
+    TODO: unify via explicit --chdir? but not sure if worth it
+
+    WARNING: Python expressions are `eval`ed within the scope of this process
+    and provided existing objects.
+
+    """
+    # TODO:     _examples_ = [], # see e.g. run
+
+    _params_ = dict(
+        cmd=Parameter(
+            args=("cmd",),
+            nargs=REMAINDER,
+            metavar='COMMAND',
+            doc="""command for execution. A leading '--' can be used to
+            disambiguate this command from the preceding options to DataLad."""),
+        python=Parameter(
+            args=("--python",),
+            action="store_true",
+            doc="""if given, must be a boolean flag indicating whether
+            the `cmd` is Python expressions to be evaluated, instead of
+            an external command to be executed"""),
+        # Following options are taken from subdatasets
+        dataset=Parameter(
+            args=("-d", "--dataset"),
+            doc="""specify the dataset to operate on.  If
+            no dataset is given, an attempt is made to identify the dataset
+            based on the input and/or the current working directory""",
+            constraints=EnsureDataset() | EnsureNone()),
+        # TODO: assume True since in 99% of use cases we just want to operate
+        #  on present subdatasets?  but having it explicit could be good to
+        #  "not miss any".  But `--fulfilled true` is getting on my nerves
```

Comment:
Hmm, yeah, I see both points but would lean toward a true default.

Reply:
If the default would not be `True`, any given command would need to check if presence of a dataset.

Reply:
https://github.com/datalad/datalad/pull/5640 is a pre-cursor to allow fulfilled to default to True here.